src/cephadm/cephadm

```diff
@@ -3450,26 +3453,159 @@ def command_list_networks():
+        dirs = os.listdir(os.path.join(data_dir, fsid))
+        for dir_name in dirs:
+            if dir_name.startswith(service):
+                if service in ['node-exporter', 'alertmanager', 'prometheus']:
+                    return _get_listen_address(os.path.join(data_dir, fsid, dir_name, 'unit.run'))
+                elif service == 'grafana':
+                    return _get_grafana_port(os.path.join(data_dir, fsid, dir_name, 'etc', 'grafana', 'grafana.ini'))
+                break
+        return None
+
+    # To get the version we're opting to use the daemon's API (if it has one), which is typically
+    # 10x faster than running a "exec <ID> <binary> --version" against the running container. We 
+    # only really use the "exec" form as a last resort to keep the runtime as low as possible
+    def _get_grafana_version(default_port: str = '3000') -> str:
+        port = _get_web_port('grafana') or default_port
+        ctx = ssl.create_default_context()
+        ctx.check_hostname = False
+        ctx.verify_mode = ssl.CERT_NONE
+        data = _http_call(
+            f"https://localhost:{port}/api/health",
+            ctx)
+        return json.loads(data)['version']
+
+    def _get_prometheus_version(default_port: str = '9090') -> str:
+        port = _get_web_port('prometheus') or default_port
+        data = _http_call(f"http://localhost:{port}/api/v1/status/buildinfo")
+        r = json.loads(data)
+        return r['data']['version']
+
+    def _get_node_exporter_version(default_port: str = '9100') -> str:
+        port = _get_web_port('node-exporter') or default_port
+        data = _http_call(f"http://localhost:{port}/metrics")
+        for metric in data.split('\n'):
+            if metric.startswith('node_exporter_build_info'):
+                # node_exporter_build_info{branch="HEAD",goversion="go1.12.5",revision="3db77732e925c08f675d7404a8c46466b2ece83e",version="0.18.1"} 1'
+                labels_str = re.split('[{}]', metric)[1]  # k=v,k=v,k=v
+                for kv_str in labels_str.split(','):
+                    k, v = kv_str.split('=')
+                    if k == 'version':
+                        return v.strip('"')
+        return ''
 
-    data_dir = args.data_dir
+    def _version_from_tag(image_name: str) -> str:
+        # if the tag looks like a version, default to that!
+        tag = image_name.split(':')[-1]
+        if re.search('^v?\d\.\d', tag):
+            return tag
+        else:
+            return ''
+    
+    def _get_tag(image_name: str) -> str:
+        return image_name.split(':')[-1]
+
+    def _load_env(env_string: str) -> Dict[str, str]:
+        env_vars = {}
+        if env_string and "=" in env_string:
+            kv_pairs = env_string[1:-1].split(' ')
+            for _i in kv_pairs:
+                k, v = _i.split('=')
+                env_vars[k] = v
+        return env_vars
+
+    version_call = {
+        "prometheus": _get_prometheus_version,
```

Comment:
Can we stay consistent to `CephIscsi` and `NFSGanesha` and use a Class based approach here?

Reply:
We don't have specific classes for the monitoring stack. We could refactor the Monitoring class later, and apply the class based approach - but I don't intend to refactor that for this PR